tests/test_player.py

```diff
@@ -0,0 +1,303 @@
+            if getattr(player, "peanut_allergy", None):
+                assert isinstance(player.peanut_allergy, bool)
+            if getattr(player, "fate", None):
+                assert isinstance(player.fate, int)
+
+    def test_simulated_copy(self, players):
+        for player in players:
+            mult = player.simulated_copy(multipliers={"overall_rating": 2.0, "patheticism": 0.01, "cinnamon": 6.9})
+            assert isinstance(mult, Player)
+            assert mult.batting_stars >= player.batting_stars
+            assert mult.patheticism < player.patheticism
+            if getattr(player, "cinnamon", None):
+                assert mult.cinnamon > player.cinnamon
+
+            buff = player.simulated_copy(buffs={"overall_rating": -2.0, "patheticism": -0.99, "cinnamon": 4.20})
+            assert isinstance(buff, Player)
+            assert buff.batting_stars <= player.batting_stars
+            assert buff.patheticism >= player.patheticism
+            if getattr(player, "cinnamon", None):
+                assert buff.cinnamon > player.cinnamon
+
+            reroll = player.simulated_copy(reroll={"overall_rating": True, "patheticism": True, "cinnamon": True})
+            assert isinstance(reroll, Player)
+
+            override = player.simulated_copy(overrides={"cinnamon": 6.9})
+            assert isinstance(override, Player)
+            assert override.cinnamon == 6.9
+
+    def test_stars_bounded(self):
+        """
+        Test that star equations produce correct results
+        """
+        with open(f"{TEST_DATA_DIR}/player/stars.json", "r") as fp:
+            players = [Player(data) for data in json.load(fp)]
+            for player in players:
+                assert player.batting_stars == pytest.approx(player.reference_batting_stars)
+                assert player.pitching_stars == pytest.approx(player.reference_pitching_stars)
+                assert player.baserunning_stars == pytest.approx(player.reference_baserunning_stars)
+                assert player.defense_stars == pytest.approx(player.reference_defense_stars)
+
+    def test_soulscream_bounded(self):
+        """
+        Test that soulscream equation produces correct results
+        """
+        with open(f"{TEST_DATA_DIR}/player/screams.json", "r") as fp:
+            players = [Player(data) for data in json.load(fp)]
+            for player in players:
+                assert player.soulscream == pytest.approx(player.reference_scream)
+
+    def test_vibes_bounded(self):
+        """
+        Test that vibe equation produces correct results
+        """
+        with open(f"{TEST_DATA_DIR}/player/vibes.json", "r") as fp:
+            players = [Player(data) for data in json.load(fp)]
+            for player in players:
+                for day in range(1, 100):
+                    assert player.get_vibe(day) == pytest.approx(player.vibes[day-1])
+
+    @pytest.mark.vcr
+    def test_blood_coffee_bounded(self):
```

Comment:
probably don't need a json file for this, just define the json in two separate tests. One thing to keep in mind is if you have a single test doing a loop of cases or a slew of assertions, you're probably undermining pytest and should convert to parametrization and/or multiple test cases.

A loop is suboptimal for two major reasons:
1) if one of the cases early in the loop fails, problems in later cases will not arise until you fix them all in series
2) mixing cases muddles the stack trace and isn't clear which input caused the problem

Reply:
So a fun problem with parametrizing this one specifically:
* In order to accurately test this, we need the error case, so we get "Coffee?" or "Blood?" as our expected result
* This test needs VCR, since the blood/coffee calls are network calls
* pytest-vcr is smart and will include the parameter name in the cassette file name
* But then Windows dies because we have a question mark in a filename....

Reply:
OK, fixed this one. Turns out `pytest-vcr` is outdated and the creator made `pytest-recording` which fixes this bug.